include/zephyr/drivers/mspi.h

```diff
@@ -0,0 +1,791 @@
+	uint32_t                ui32CENum;
+	/** @brief Configure frequency */
+	uint32_t                ui32Freq;
+	/** @brief Configure I/O mode */
+	enum mspi_io_mode       eIOMode;
+	/** @brief Configure data rate */
+	enum mspi_data_rate     eDataRate;
+	/** @brief Configure clock polarity and phase*/
+	enum mspi_cpp_mode      eCPP;
+	/** @brief Configure transfer endian */
+	enum mspi_endian        eEndian;
+	/** @brief Configure chip enable polarity */
+	enum mspi_ce_polarity   eCEPolarity;
+	/** @brief Configure DQS mode */
+	bool                    bDQSEnable;
+	/** @brief Configure number of clock cycles between
+	 * addr and data in RX direction
+	 */
+	uint32_t                ui32RXDummy;
+	/** @brief Configure number of clock cycles between
+	 * addr and data in TX direction
+	 */
+	uint32_t                ui32TXDummy;
+	/** @brief Configure read instruction */
+	uint32_t                ui32ReadInstr;
+	/** @brief Configure write instruction */
+	uint32_t                ui32WriteInstr;
+	/** @brief Configure instruction length */
+	uint16_t                ui16InstrLength;
+	/** @brief Configure address length */
+	uint16_t                ui16AddrLength;
+	/** @brief Configure memory boundary */
+	uint32_t                ui32MemBoundary;
+	/** @brief Configure break time */
+	uint32_t                ui32BreakTimeLimit;
+};
+
+/**
+ * @brief MSPI controller XIP configuration
+ */
+struct mspi_xip_cfg {
+	/** @brief XIP enable */
+	bool                    bEnable;
+	/** @brief XIP region start address =
+	 * hardware default + address offset
+	 */
+	uint32_t                ui32AddrOffset;
+	/** @brief XIP region size */
+	uint32_t                ui32Size;
+	/** @brief XIP access permission */
+	enum mspi_xip_permit    ePermission;
+};
+
+/**
+ * @brief MSPI controller scramble configuration
+ */
+struct mspi_scramble_cfg {
+	/** @brief scramble enable */
+	bool                    bEnable;
+	/** @brief scramble region start address =
+	 * hardware default + address offset
+	 */
+	uint32_t                ui32AddrOffset;
+	/** @brief scramble region size */
+	uint32_t                ui32Size;
+};
+
+/** @} */
+
+/**
+ * @brief MSPI Transfer API
+ * @defgroup mspi_transfer_api MSPI Transfer API
+ * @{
+ */
+
+/**
+ * @brief MSPI Chip Select control structure
+ *
+ * This can be used to control a CE line via a GPIO line, instead of
+ * using the controller inner CE logic.
+ *
+ */
+struct mspi_ce_control {
+	/**
+	 * @brief GPIO devicetree specification of CE GPIO.
+	 * The device pointer can be set to NULL to fully inhibit CE control if
+	 * necessary. The GPIO flags GPIO_ACTIVE_LOW/GPIO_ACTIVE_HIGH should be
+	 * the same as in MSPI configuration.
+	 */
+	struct gpio_dt_spec gpio;
+	/**
+	 * @brief Delay to wait.
+	 * In microseconds before starting the
+	 * transmission and before releasing the CE line.
+	 */
+	uint32_t            delay;
+};
+
+/**
+ * @brief MSPI peripheral xfer packet format
+ */
+struct mspi_xfer_packet {
```

Comment:
I believe each packet should support setting the number of pads to use- this will allow us to better support cases where a MSPI device might support 1-4-4 reads, for example

Reply:
As I indicated in other comment, switching between op-modes may require quite some time for both controller and devices. The mspi_xfer_packet is intended for bulk transfer and scatter IO operation not combining register access and memory access which may have different op-mode requirements.

Reply:
> As I indicated in other comment, switching between op-modes may require quite some time for both controller and devices

But what would the disadvantage be of combining a register and memory access command into one `mspi_transceive` call? If the controller hardware knows it will be faster to "chunk" the transfers based on the pad count used, then I feel that logic can be implemented within the controller driver, and should not leak out into the interface that MSPI devices use.

Reply:
The way I see it is that the list of settings would explode and eventually cover everything in `mspi_dev_cfg`.
You do realize that register access may only happened at lower frequencies? Then should we include frequency into `mspi_xfer`? An example device is ATXP032, the memory access frequencies can be higher than register access.
What we need is an API that balances between all these different controller and devices. Surely I value your input, but I also thinks that **we need to consider real world usage**. In general, for memory devices, you would have 90% of the memory access during run-time. For PSRAM, that number can go up to 99% as you would only need it for initialization and sleep.  For other SPI devices, such quirk back and forth with the io-mode is even rare as these quirks are really the result of memory manufaturers trying to bump memory performance. 

As I indicated in my other comment, the boundraies of APIs become blury when you have different APIs that could do the same thing. I prefer that a lower level API is clear and concise.

Reply:
To be honst, I see the list of settings in `mspi_xfer` starts to grow uncontrollably and we need discuss it with a base line for reference, something like JESD251C.